src/stan/interface_callbacks/writer/stream_writer.hpp

```diff
@@ -0,0 +1,101 @@
+#ifndef STAN_INTERFACE_CALLBACKS_WRITER_STREAM_WRITER_HPP
+#define STAN_INTERFACE_CALLBACKS_WRITER_STREAM_WRITER_HPP
+
+#include <stan/interface_callbacks/writer/base_writer.hpp>
+#include <iostream>
+#include <vector>
+#include <string>
+
+namespace stan {
+  namespace interface_callbacks {
+    namespace writer {
+
+      template <class Stream>
+      class stream_writer: public base_writer {
+      public:
+        stream_writer(Stream& output,
+                      const std::string& key_value_prefix = ""):
+          output__(output), key_value_prefix__(key_value_prefix) {}
+
+        void operator()(const std::string& key, double value) {
+          output__ << key_value_prefix__ << key << " = " << value << std::endl;
+        }
+
+        void operator()(const std::string& key, int value) {
+          output__ << key_value_prefix__ << key << " = " << value << std::endl;
+        }
+
+        void operator()(const std::string& key, const std::string& value) {
+          output__ << key_value_prefix__ << key << " = " << value << std::endl;
+        }
+
+        void operator()(const std::string& key,
+                        const double* values,
+                        int n_values) {
+          if (n_values == 0) return;
+
+          output__ << key_value_prefix__ << key << ": ";
+
+          output__ << values[0];
+          for (int n = 1; n < n_values; ++n)
+            output__ << "," << values[n];
+          output__ << std::endl;
+        }
+
+        void operator()(const std::string& key,
+                        const double* values,
+                        int n_rows, int n_cols) {
+          if (n_rows == 0 || n_cols == 0) return;
+
+          output__ << key_value_prefix__ << key << std::endl;
+
+          for (int i = 0; i < n_rows; ++i) {
+            output__ << key_value_prefix__ << values[i * n_cols];
+            for (int j = 1; j < n_cols; ++j)
+              output__ << "," << values[i * n_cols + j];
+            output__ << std::endl;
+          }
+        }
+
+        void operator()(const std::vector<std::string>& names) {
+          if (names.empty()) return;
+
+          std::vector<std::string>::const_iterator last = names.end();
+          --last;
+
+          for (std::vector<std::string>::const_iterator it = names.begin();
+               it != last; ++it)
+            output__ << *it << ",";
+          output__ << names.back() << std::endl;
+        }
+
+        void operator()(const std::vector<double>& state) {
+          if (state.empty()) return;
+
+          std::vector<double>::const_iterator last = state.end();
+          --last;
+
+          for (std::vector<double>::const_iterator it = state.begin();
+               it != last; ++it)
+            output__ << *it << ",";
+          output__ << state.back() << std::endl;
+        }
+
+        void operator()() {
+          output__ << std::endl;
```

Comment:
It seems more consistent that way. As it stands, all the methods for writing have comments except for:
1. writing the names (header of csv)
2. writing the state
3. writing a blank line.
The third seems out of place when the other methods like writing a message and writing key / value pairs include the `key_value_prefix__`. (wow... I'm just realizing what the variable name is indicating.)
If we were to go the other way, I would much rather not print the `key_value_prefix__` all method calls that aren't for key / value pairs. I think that might change behavior, though.